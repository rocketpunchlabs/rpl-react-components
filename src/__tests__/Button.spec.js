import React from 'react';
import styled from 'styled-components';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';
import { shallow, mount, render, configure } from 'enzyme';
import renderer from 'react-test-renderer';
import { Button } from '../';

// setup file
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

// test('it works', () => {
//   const tree = renderer.create(<Button>Hello</Button>).toJSON();
//   expect(tree).toMatchSnapshot();
//   console.log(tree);
//   expect(tree).toHaveStyleRule('color', 'black!important');
// });

describe('Button', () => {
  it('renders without crashing', () => {
    const tree = renderer.create(<Button>Hello</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should default to black text', () => {
    const tree = renderer.create(<Button>Hello</Button>).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('color', 'black!important');
  });

  it('should default to white background', () => {
    const tree = renderer.create(<Button>Hello</Button>).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('background', 'white');
  });

  it('should render children', () => {
    const wrapper = mount(<Button>Hello World</Button>);
    expect(wrapper.find('a').text()).toBe('Hello World');
  });

  it('should render large button', () => {
    const tree = renderer.create(<Button large>Hello</Button>).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('font-size', '24px');
  });

  it('should render small button', () => {
    const tree = renderer.create(<Button small>Hello</Button>).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('font-size', '12px');
  });

  it('should render blue button', () => {
    const tree = renderer
      .create(<Button bgColor="#3498db">Hello</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('background', '#3498db');
  });

  it('should render block button', () => {
    const tree = renderer.create(<Button block>Hello</Button>).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('display', 'block');
  });

  it('should render a link', () => {
    const tree = renderer
      .create(<Button href="https://www.google.com">Google</Button>)
      .toJSON();
    const wrapper = mount(
      <Button href="https://www.google.com">Google</Button>
    );
    expect(tree).toMatchSnapshot();
    expect(wrapper.find('a').prop('href')).toBe('https://www.google.com');
  });

  it('should pass an onClick', () => {
    const tree = renderer
      .create(<Button onClick={() => alert('clicked')}>Hello</Button>)
      .toJSON();
    const wrapper = mount(
      <Button onClick={() => console.log('clicked')}>Hello</Button>
    );
    expect(tree).toMatchSnapshot();
  });

  it('should perform the onClick', () => {
    const tree = renderer
      .create(<Button onClick={() => console.log('clicked')}>Hello</Button>)
      .toJSON();
    // const wrapper = mount(
    //   <Button onClick={() => console.log('clicked')}>Hello</Button>
    // );
    expect(tree).toMatchSnapshot();

    tree.props.onClick();
    expect(tree).toMatchSnapshot();
  });

  //   it('should have "large" prop', () => {
  //     const wrapper = shallow(<Button large>Large</Button>);
  //     expect(wrapper.prop('large')).toBe(true);
  //   });

  //   it('should have "small" prop', () => {
  //     const wrapper = shallow(<Button small>Small</Button>);
  //     expect(wrapper.prop('small')).toBe(true);
  //   });

  //   it('should have "bgColor" prop', () => {
  //     const wrapper = shallow(<Button bgColor="#ec12ca">Small</Button>);
  //     expect(wrapper.prop('bgColor')).toEqual('#ec12ca');
  //   });

  //   it('should have style', () => {
  //     const wrapper = shallow(<Button bgColor="#ec12ca">Small</Button>);
  //     console.log(wrapper);
  //     expect(wrapper).to.have.style('border');
  //     // expect(wrapper.prop('bgColor')).toEqual('#ec12ca');
  //   });

  //   it('should render an anchor element if href exists', () => {
  //     const wrapper = mount(<Button href="/home">Home</Button>);

  //     expect(wrapper.find('a').length).toBe(1);
  //     expect(wrapper.text()).toBe('Home');
  //   });

  //   it('should render type as undefined by default when tag is "button"', () => {
  //     const wrapper = mount(<Button>Home</Button>);

  //     expect(wrapper.find('button').prop('type')).toBe(undefined);
  //     expect(wrapper.text()).toBe('Home');
  //   });

  //   it('should render type as "button" by default when tag is "button" and onClick is provided', () => {
  //     const wrapper = mount(<Button onClick={() => {}}>Home</Button>);

  //     expect(wrapper.find('button').prop('type')).toBe('button');
  //     expect(wrapper.text()).toBe('Home');
  //   });

  //   it('should render type as user defined when defined by the user', () => {
  //     const wrapper = mount(<Button type="submit">Home</Button>);

  //     expect(wrapper.find('button').prop('type')).toBe('submit');
  //     expect(wrapper.text()).toBe('Home');
  //   });

  //   it('should not render type by default when the type is not defined and the tag is not "button"', () => {
  //     const wrapper = mount(<Button tag="a">Home</Button>);

  //     expect(wrapper.find('a').prop('type')).toBe(undefined);
  //     expect(wrapper.text()).toBe('Home');
  //   });

  //   it('should not render type by default when the type is not defined and the href is defined', () => {
  //     const wrapper = mount(<Button href="#">Home</Button>);

  //     expect(wrapper.find('a').prop('type')).toBe(undefined);
  //     expect(wrapper.text()).toBe('Home');
  //   });

  //   it('should render buttons with default color', () => {
  //     const wrapper = shallow(<Button>Default Button</Button>);

  //     expect(wrapper.hasClass('btn-secondary')).toBe(true);
  //   });

  //   it('should render buttons with other colors', () => {
  //     const wrapper = shallow(<Button color="danger">Default Button</Button>);

  //     expect(wrapper.hasClass('btn-danger')).toBe(true);
  //   });

  //   it('should render buttons with outline variant', () => {
  //     const wrapper = shallow(<Button outline>Default Button</Button>);

  //     expect(wrapper.hasClass('btn-outline-secondary')).toBe(true);
  //   });

  //   it('should render buttons with outline variant with different colors', () => {
  //     const wrapper = shallow(<Button outline color="info">Default Button</Button>);

  //     expect(wrapper.hasClass('btn-outline-info')).toBe(true);
  //   });

  //   it('should render buttons at different sizes', () => {
  //     const small = shallow(<Button size="sm">Small Button</Button>);
  //     const large = shallow(<Button size="lg">Large Button</Button>);

  //     expect(small.hasClass('btn-sm')).toBe(true);
  //     expect(large.hasClass('btn-lg')).toBe(true);
  //   });

  //   it('should render block level buttons', () => {
  //     const block = shallow(<Button block>Block Level Button</Button>);

  //     expect(block.hasClass('btn-block')).toBe(true);
  //   });

  //   describe('onClick', () => {
  //     it('calls props.onClick if it exists', () => {
  //       const onClick = jest.fn();
  //       const wrapper = mount(<Button onClick={onClick}>Testing Click</Button>);

  //       wrapper.find('button').simulate('click');
  //       expect(onClick).toHaveBeenCalled();
  //     });

  //     it('is not called when disabled', () => {
  //       const e = createSpyObj('e', ['preventDefault']);
  //       const wrapper = mount(<Button>Testing Click</Button>);

  //       wrapper.instance().onClick(e);
  //       expect(e.preventDefault).not.toHaveBeenCalled();

  //       wrapper.setProps({ disabled: true });
  //       wrapper.instance().onClick(e);
  //       expect(e.preventDefault).toHaveBeenCalled();
  //     });
  //   });
});
