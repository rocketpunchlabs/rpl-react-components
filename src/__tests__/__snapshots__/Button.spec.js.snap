// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders without crashing 1`] = `
.c0 {
  background: white;
  font-family: sans-serif;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  display: inline-block;
  font-size: 15px;
  font-weight: 400;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  color: black!important;
  border-radius: 4px;
  padding: 9px 24px;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
}

.c0:hover {
  text-decoration: none;
  background: hsl(0,0%,90%);
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  box-shadow: 0 7px 14px rgba(50,50,93,0.1),0 3px 6px rgba(0,0,0,0.08);
}

<a
  className="c0"
>
  Hello
</a>
`;

exports[`Button should default to black text 1`] = `
.c0 {
  background: white;
  font-family: sans-serif;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  display: inline-block;
  font-size: 15px;
  font-weight: 400;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  color: black!important;
  border-radius: 4px;
  padding: 9px 24px;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
}

.c0:hover {
  text-decoration: none;
  background: hsl(0,0%,90%);
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  box-shadow: 0 7px 14px rgba(50,50,93,0.1),0 3px 6px rgba(0,0,0,0.08);
}

<a
  className="c0"
>
  Hello
</a>
`;

exports[`Button should default to white background 1`] = `
.c0 {
  background: white;
  font-family: sans-serif;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  display: inline-block;
  font-size: 15px;
  font-weight: 400;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  color: black!important;
  border-radius: 4px;
  padding: 9px 24px;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
}

.c0:hover {
  text-decoration: none;
  background: hsl(0,0%,90%);
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  box-shadow: 0 7px 14px rgba(50,50,93,0.1),0 3px 6px rgba(0,0,0,0.08);
}

<a
  className="c0"
>
  Hello
</a>
`;

exports[`Button should pass an onClick 1`] = `
.c0 {
  background: white;
  font-family: sans-serif;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  display: inline-block;
  font-size: 15px;
  font-weight: 400;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  color: black!important;
  border-radius: 4px;
  padding: 9px 24px;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
}

.c0:hover {
  text-decoration: none;
  background: hsl(0,0%,90%);
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  box-shadow: 0 7px 14px rgba(50,50,93,0.1),0 3px 6px rgba(0,0,0,0.08);
}

<a
  className="c0"
  onClick={[Function]}
>
  Hello
</a>
`;

exports[`Button should perform the onClick 1`] = `
.c0 {
  background: white;
  font-family: sans-serif;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  display: inline-block;
  font-size: 15px;
  font-weight: 400;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  color: black!important;
  border-radius: 4px;
  padding: 9px 24px;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
}

.c0:hover {
  text-decoration: none;
  background: hsl(0,0%,90%);
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  box-shadow: 0 7px 14px rgba(50,50,93,0.1),0 3px 6px rgba(0,0,0,0.08);
}

<a
  className="c0"
  onClick={[Function]}
>
  Hello
</a>
`;

exports[`Button should perform the onClick 2`] = `
<a
  className="sc-bdVaJa ljxPpA"
  onClick={[Function]}
>
  Hello
</a>
`;

exports[`Button should render a link 1`] = `
.c0 {
  background: white;
  font-family: sans-serif;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  display: inline-block;
  font-size: 15px;
  font-weight: 400;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  color: black!important;
  border-radius: 4px;
  padding: 9px 24px;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
}

.c0:hover {
  text-decoration: none;
  background: hsl(0,0%,90%);
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  box-shadow: 0 7px 14px rgba(50,50,93,0.1),0 3px 6px rgba(0,0,0,0.08);
}

<a
  className="c0"
  href="https://www.google.com"
>
  Google
</a>
`;

exports[`Button should render block button 1`] = `
.c0 {
  background: white;
  font-family: sans-serif;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  display: block;
  font-size: 15px;
  font-weight: 400;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  color: black!important;
  border-radius: 4px;
  padding: 9px 24px;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
}

.c0:hover {
  text-decoration: none;
  background: hsl(0,0%,90%);
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  box-shadow: 0 7px 14px rgba(50,50,93,0.1),0 3px 6px rgba(0,0,0,0.08);
}

<a
  className="c0"
>
  Hello
</a>
`;

exports[`Button should render blue button 1`] = `
.c0 {
  background: #3498db;
  font-family: sans-serif;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  display: inline-block;
  font-size: 15px;
  font-weight: 400;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  color: white!important;
  border-radius: 4px;
  padding: 9px 24px;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
}

.c0:hover {
  text-decoration: none;
  background: hsl(204.10000000000002,69.9%,47.8%);
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  box-shadow: 0 7px 14px rgba(50,50,93,0.1),0 3px 6px rgba(0,0,0,0.08);
}

<a
  className="c0"
>
  Hello
</a>
`;

exports[`Button should render large button 1`] = `
.c0 {
  background: white;
  font-family: sans-serif;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  display: inline-block;
  font-size: 15px;
  font-weight: 400;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  color: black!important;
  border-radius: 4px;
  padding: 9px 24px;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
  font-size: 24px;
  line-height: 24px;
}

.c0:hover {
  text-decoration: none;
  background: hsl(0,0%,90%);
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  box-shadow: 0 7px 14px rgba(50,50,93,0.1),0 3px 6px rgba(0,0,0,0.08);
}

<a
  className="c0"
>
  Hello
</a>
`;

exports[`Button should render small button 1`] = `
.c0 {
  background: white;
  font-family: sans-serif;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  display: inline-block;
  font-size: 15px;
  font-weight: 400;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  color: black!important;
  border-radius: 4px;
  padding: 9px 24px;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
  font-size: 12px;
  padding: 6px 18px;
}

.c0:hover {
  text-decoration: none;
  background: hsl(0,0%,90%);
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  box-shadow: 0 7px 14px rgba(50,50,93,0.1),0 3px 6px rgba(0,0,0,0.08);
}

<a
  className="c0"
>
  Hello
</a>
`;
